test/stdlib/TracePoint_test.rb

```diff
@@ -1,160 +1,233 @@
+require_relative 'test_helper'
 
 class TracePointSingletonTest < Test::Unit::TestCase
   include TestHelper
 
-  testing "singleton(::TracePoint)"
+  testing 'singleton(::TracePoint)'
+
+  MyTracePoint = Class.new(TracePoint)
 
   def test_new
-    assert_send_type  "(*::Symbol events) { (::TracePoint tp) -> void } -> ::TracePoint",
-                      TracePoint, :new, :line do end
+    assert_send_type  '(*::_ToSym) { (TracePointSingletonTest::MyTracePoint tp) -> void } -> TracePointSingletonTest::MyTracePoint',
+                      MyTracePoint, :new, :line, ToSym.new(:line) do end
   end
 
-  # TODO
-  # def test_allow_reentry
-  #   assert_send_type  "() { () -> void } -> void",
-  #                     TracePoint, :allow_reentry
-  # end
+  def test_allow_reentry
+    tp = TracePoint.new :script_compiled do
+      assert_send_type  '[T] () { (nil) -> T } -> T',
+                        TracePoint, :allow_reentry do 1r end
+    end
+    tp.enable
+
+    eval("1") # some no-op to trigger the tracepoint
+  ensure
+    tp.disable
+  end
 
   def test_stat
-    assert_send_type  "() -> untyped",
+    assert_send_type  '() -> untyped',
                       TracePoint, :stat
   end
 
   def test_trace
-    assert_send_type  "(*::Symbol events) { (::TracePoint tp) -> void } -> ::TracePoint",
-                      TracePoint, :trace, :line do |tp| tp.disable end
+    # Make sure to pick a tracepoint that won't be accidentaly called, to prevent loops.
+    tp = assert_send_type  '(*::_ToSym) { (TracePointSingletonTest::MyTracePoint tp) -> void } -> TracePointSingletonTest::MyTracePoint',
+                           MyTracePoint, :trace, :script_compiled, ToSym.new(:script_compiled) do end
+    tp.disable
   end
 end
 
 class TracePointTest < Test::Unit::TestCase
   include TestHelper
 
-  testing "::TracePoint"
+  testing '::TracePoint'
 
-  def test_initialize
-    assert_send_type  "(*::Symbol events) { (::TracePoint tp) -> void } -> void",
-                      TracePoint.new(:line){}, :initialize do end
+  def tracepoint(event=:line, do: proc{ 1 }, &tp_body)
+    tp = TracePoint.new(event, &tp_body)
+    tp.enable(&binding.local_variable_get(:do))
+  ensure
+    tp.disable
   end
-
+  
   def test_binding
-    TracePoint.new(:line) do |tp|
-      assert_send_type "() -> ::Binding",
-                       tp, :binding
-    end.enable { 1 }
-  end
+    tracepoint do |tp|
+      assert_send_type  '() -> Binding',
+                        tp, :binding
+    end
 
-  def test_inspect
-    TracePoint.new(:line) do |tp|
-      assert_send_type "() -> ::String",
-                       tp, :inspect
-    end.enable { 1 }
+    tracepoint :c_call, do: method(:print) do |tp|
+      assert_send_type  '() -> nil',
+                        tp, :binding
+    end
   end
 
+  TracePoint.new(:b_return) { |tp|
+    const_set :CALLEE_ID_OUTSIDE_OF_A_METHOD, tp.callee_id
+    const_set :METHOD_ID_OUTSIDE_OF_A_METHOD, tp.method_id
+  }.enable { }
```

Comment:
There's no way to get `nil` for the  `callee_id` and `method_id` methods within a method—it needs to be done at top-level.